tiledb/sm/array_schema/dimension.cc

```diff
@@ -59,6 +60,193 @@ class DimensionException : public StatusException {
   }
 };
 
+/* ************************ */
+/*     DYNAMIC DISPATCH     */
+/* ************************ */
+
+/**
+ * Dispatches Dimension behavior based on the physical type
+ * of the Dimension
+ *
+ * Subclasses DimensionDispatchFixedSize and DimensionDispatchVarSize
+ * handle the few scenarios where the dispatch is based on
+ * something beyond just the type (namely, function arguments).
+ */
+template <class T>
+class DimensionDispatchTyped : public Dimension::DimensionDispatch {
+ public:
+  using DimensionDispatch::DimensionDispatch;
+
+ protected:
+  void ceil_to_tile(
+      const Range& r, uint64_t tile_num, ByteVecValue* v) const override {
+    return Dimension::ceil_to_tile<T>(&base, r, tile_num, v);
+  }
+  bool check_range(const Range& range, std::string* error) const override {
+    return Dimension::check_range<T>(&base, range, error);
+  }
+  bool coincides_with_tiles(const Range& r) const override {
+    return Dimension::coincides_with_tiles<T>(&base, r);
+  }
+  void crop_range(Range* range) const override {
+    return Dimension::crop_range<T>(&base, range);
+  }
+  uint64_t domain_range(const Range& range) const override {
+    return Dimension::domain_range<T>(range);
+  }
+  void expand_range(const Range& r1, Range* r2) const override {
+    return Dimension::expand_range<T>(r1, r2);
+  }
+  void expand_range_v(const void* v, Range* r) const override {
+    return Dimension::expand_range_v<T>(v, r);
+  }
+  void expand_to_tile(Range* range) const override {
+    return Dimension::expand_to_tile<T>(&base, range);
+  }
+  bool oob(const void* coord, std::string* err_msg) const override {
+    return Dimension::oob<T>(&base, coord, err_msg);
+  }
+  bool covered(const Range& r1, const Range& r2) const override {
+    if constexpr (std::is_same_v<T, char>) {
+      assert(base.var_size());
+    }
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::covered<T>(r1, r2);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  bool overlap(const Range& r1, const Range& r2) const override {
+    if constexpr (std::is_same_v<T, char>) {
+      assert(base.var_size());
+    }
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::overlap<T>(r1, r2);
+    } else {
+      ::stdx::unreachable();
```

Comment:
We should probably turn this into a `static_assert(tiledb::type::TileDBFundamental<T>)` or whatever. R has a fairly strict stance on extension modules not being allowed to terminate the process so we using things like unreachable mean we should only be putting them places that are visually inspectable to be true.

Reply:
And/or places that'll trigger in CI for use in things like demonstrating switch cases cover all enum members or the like.

Reply:
Ah, we for sure want these to be static_assert's now that I think about it. They'll fail horribly in CI if these aren't but we'd rather have them fail at compile time.

Reply:
I was skeptical that `static_assert` would actually work here since it is overriding a virtual function - but I tried it out and it does. Nice surprise!
I'm not sure how the compiler can tell that the method is never called for certain template instantiations but I guess it can.